Chromium Chronicle 25 (#1493)
GoogleChrome/developer.chrome.com: site/en/blog/chromium-chronicle-25/index.md

---
title: "The Chromium Chronicle #25: Thread Safety Annotations"
description: >
  Learn how Clang's static analysis framework takes over the toil of thread-safety proofs.
layout: 'layouts/blog-post.njk'

> date: 2021-09-27

Author: ```suggestion
date: 2021-10-01
```

> date: 2021-10-01

Author: ```suggestion
date: 2021-10-04
```

hero: 'image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/hgu6uTktp2ipmuODZZhP.jpg'
alt: >
  Chromium Chronicle image
tags:
  - chromium-chronicle
---

> **Episode 25:** by Victor Costan in SFO (September, 2021)<br>

Reviewer: Thanks for making all the edits, Rachel! Could we please update this time frame "(September, 2021)" to instead be "(October, 2021)", since we didn't get to push last week?

Author: ```suggestion
**Episode 25:** by Victor Costan in SFO (October, 2021)<br>
```

> **Episode 25:** by Victor Constan in SFO (September, 2021)<br>

Reviewer 2: Constan -> Costan

(sorry for the late comment, I just noticed this)

Author: ```suggestion
**Episode 25:** by Victor Costan in SFO (September, 2021)<br>
```

[Previous episodes](/tags/chromium-chronicle/)

In C++, ruling out the possibility of data races comes down to a small thread-safety correctness proof for every data member access.
These proofs add up to a lot of mental toil, especially when reviewing or refactoring code.
**Clang's static analysis framework takes over the toil of thread-safety proofs**.

## Add `GUARDED_BY_CONTEXT()` to data members in thread-unsafe classes

Most Chrome classes are thread-unsafe, and should be used on a single sequence.

> Add annotations to all data members that are not thread-safe.

Reviewer 3: Practically, we're not going to annotate all members in all classes.  Consider guidance for when a class should have annotations.  e.g. "Annotate classes that are used in multithreading circumstances.  Add annotations to all data members..."

Reviewer 2: I'd prefer that we keep this TotT short and sweet. 

I don't think we have consensus yet on what would get annotated, and I'd like to get this nudge out to folks as soon as possible. I am asking folks to [annotate *all* classes in storage code](https://docs.google.com/document/d/1QRIPkrOKU-QdDOpT4nak9UfMT8SKAzZfbe13fBah7ow/edit#heading=h.8ffdyerzmckm).

Author: Do we need to change anything based on this comment?

Reviewer 3: The article does not need to change, but I stand by the suggestion.  I think readers will be confused when to do this.

Unnecessary annotations are safe, but missing annotations are a risk of data races.

```cpp
#include "base/sequence_checker.h"  // for SEQUENCE_CHECKER()
#include "base/thread_annotations.h"  // for GUARDED_BY_CONTEXT()

class Cache {
  // Methods here.
 private:
  SEQUENCE_CHECKER(sequence_checker_);
  base::flat_map<std::string, std::string> data_ GUARDED_BY_CONTEXT(sequence_checker_);
};
```

## Clang enforces sequence checks

In return for annotating the data members,
Clang ensures that any method that accesses the data performs a sequence safety check before doing so.
As code is moved around in refactorings,
Clang continues enforcing the `GUARDED_BY_CONTEXT()` annotation.

```cpp
void Cache::Set(base::StringPiece key, base::StringPiece value) {

>   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);  // Clang errors out without this.

Reviewer 2: (from internal draft review)

Clang errors out -> Clang warns

Author: ```suggestion
  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);  // Clang warns without this.
```

  data_.emplace(key, value);
}
```

## Add `GUARDED_BY()` to data members in thread-safe classes that use mutexes

Some classes in Chrome must use locks for thread safety.
In these cases, annotate all data members that are not thread-safe.
Each annotation points to a mutex that must be held while accessing the data member.

```cpp
#include "base/thread_annotations.h"  // for GUARDED_BY()

class ThreadSafeCache {
  // Methods here.
  private:
    base::Lock lock_;
    base::flat_map<std::string, std::string> data_ GUARDED_BY(lock_);
};
```

## Clang enforces lock acquisitions

Hang back and let the compiler ensure that each `base::AutoLock` is correctly scoped,

> and that lock `Acquire(`) and `Release()` calls are paired correctly.

Reviewer: Can we please include the closing parentheses ")" within the `` characters?

Author: ```suggestion
and that lock `Acquire()` and `Release()` calls are paired correctly.
```

```cpp
void ThreadSafeCache::Set(base::StringPiece key, base::StringPiece value) {

>   base::AutoLock auto_lock(lock_);  // Clang errors out without this.

Reviewer 2: (from internal draft review)

Clang errors out -> Clang warns

Author: ```suggestion
  base::AutoLock auto_lock(lock_);  // Clang warns without this.
```

  data_.emplace(key, value);
}
```

> - [Threading in Chrome](https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/docs/threading_and_tasks.md)

Reviewer 4: ```suggestion
- [Threading and Tasks in Chrome](https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/docs/threading_and_tasks.md)
```

> - [Clang Thread Safety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html.): Learn about other Clang annotations for more complex scenarios.

Reviewer 4: ```suggestion
- [Clang Thread Safety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html): Learn about other Clang annotations for more complex scenarios.
```
